pennylane/transforms/condition.py

```diff
@@ -233,6 +306,24 @@ def qnode(a, x, y, z):
             >>> qnode(par, x, y, z)
             tensor(-0.30922805, requires_grad=True)
     """
+    if active_jit := compiler.active_compiler():
+        available_eps = compiler.AvailableCompilers.names_entrypoints
+        ops_loader = available_eps[active_jit]["ops"].load()
+        cond_func = ops_loader.cond(condition)(true_fn)
+
+        # Optional 'elif' branches
+        for cond_val, elif_fn in elifs:
+            cond_func.else_if(cond_val)(elif_fn)

```

Comment:
I see it isn't the case, but I suppose this logic is hidden within catalyst itself? feels weird to have to keep calling every `else_if` even if the first `cond_val` was true

Reply:
When compiling (or tracing), each branch does indeed need to be "executed", but only symbolically (traced). The code in each branch is captured in the program that way, and which one is executed will be determined at runtime.

Reply:
This is the way that your program will be compiled including all of those elif branches, so that if at runtime (within next calls), these cond_vals can be changed without necessarily re-compiling.

Consider this example:

``` python
        @qml.qjit
        @qml.qnode(dev)
        def circuit(x):
            def true_fn():
                qml.RX(x, wires=0)

            def elif1_fn():
                qml.RY(x, wires=0)

            def elif2_fn():
                qml.RZ(x, wires=0)

            def false_fn():
                qml.RX(x**2, wires=0)

            qml.cond(x > 2.7, true_fn, false_fn, ((x > 2.4, elif1_fn), (x > 1.4, elif2_fn)))

            return qml.expval(qml.PauliZ(0))
```

It'll be compiled down to multiple if-else calls,

```
  func.func private @circuit(%arg0: tensor<f64>) -> tensor<f64> attributes {diff_method = "parameter-shift", llvm.linkage = #llvm.linkage<internal>, qnode} {
    %0 = stablehlo.constant dense<2.700000e+00> : tensor<f64>
    %1 = stablehlo.compare  GT, %arg0, %0,  FLOAT : (tensor<f64>, tensor<f64>) -> tensor<i1>
    %2 = stablehlo.constant dense<2.400000e+00> : tensor<f64>
    %3 = stablehlo.compare  GT, %arg0, %2,  FLOAT : (tensor<f64>, tensor<f64>) -> tensor<i1>
    %4 = stablehlo.constant dense<1.400000e+00> : tensor<f64>
    %5 = stablehlo.compare  GT, %arg0, %4,  FLOAT : (tensor<f64>, tensor<f64>) -> tensor<i1>
    quantum.device ["kwargs", "{'shots': 0}"]
    quantum.device ["backend", "lightning.qubit"]
    %6 = stablehlo.constant dense<1> : tensor<i64>
    %7 = quantum.alloc( 1) : !quantum.reg
    %extracted = tensor.extract %1[] : tensor<i1>
    %8 = scf.if %extracted -> (!quantum.reg) {
      %13 = stablehlo.constant dense<0> : tensor<i64>
      %extracted_1 = tensor.extract %13[] : tensor<i64>
      %14 = quantum.extract %7[%extracted_1] : !quantum.reg -> !quantum.bit
      %extracted_2 = tensor.extract %arg0[] : tensor<f64>
      %15 = quantum.custom "RX"(%extracted_2) %14 : !quantum.bit
      %16 = stablehlo.constant dense<0> : tensor<i64>
      %extracted_3 = tensor.extract %16[] : tensor<i64>
      %17 = quantum.insert %7[%extracted_3], %15 : !quantum.reg, !quantum.bit
      scf.yield %17 : !quantum.reg
    } else {
      %extracted_1 = tensor.extract %3[] : tensor<i1>
      %13 = scf.if %extracted_1 -> (!quantum.reg) {
        %14 = stablehlo.constant dense<0> : tensor<i64>
        %extracted_2 = tensor.extract %14[] : tensor<i64>
        %15 = quantum.extract %7[%extracted_2] : !quantum.reg -> !quantum.bit
        %extracted_3 = tensor.extract %arg0[] : tensor<f64>
        %16 = quantum.custom "RY"(%extracted_3) %15 : !quantum.bit
        %17 = stablehlo.constant dense<0> : tensor<i64>
        %extracted_4 = tensor.extract %17[] : tensor<i64>
        %18 = quantum.insert %7[%extracted_4], %16 : !quantum.reg, !quantum.bit
        scf.yield %18 : !quantum.reg
      } else {
        %extracted_2 = tensor.extract %5[] : tensor<i1>
        %14 = scf.if %extracted_2 -> (!quantum.reg) {
          %15 = stablehlo.constant dense<0> : tensor<i64>
          %extracted_3 = tensor.extract %15[] : tensor<i64>
          %16 = quantum.extract %7[%extracted_3] : !quantum.reg -> !quantum.bit
          %extracted_4 = tensor.extract %arg0[] : tensor<f64>
          %17 = quantum.custom "RZ"(%extracted_4) %16 : !quantum.bit
          %18 = stablehlo.constant dense<0> : tensor<i64>
          %extracted_5 = tensor.extract %18[] : tensor<i64>
          %19 = quantum.insert %7[%extracted_5], %17 : !quantum.reg, !quantum.bit
          scf.yield %19 : !quantum.reg
        } else {
          %15 = stablehlo.multiply %arg0, %arg0 : tensor<f64>
          %16 = stablehlo.constant dense<0> : tensor<i64>
          %extracted_3 = tensor.extract %16[] : tensor<i64>
          %17 = quantum.extract %7[%extracted_3] : !quantum.reg -> !quantum.bit
          %extracted_4 = tensor.extract %15[] : tensor<f64>
          %18 = quantum.custom "RX"(%extracted_4) %17 : !quantum.bit
          %19 = stablehlo.constant dense<0> : tensor<i64>
          %extracted_5 = tensor.extract %19[] : tensor<i64>
          %20 = quantum.insert %7[%extracted_5], %18 : !quantum.reg, !quantum.bit
          scf.yield %20 : !quantum.reg
        }
        scf.yield %14 : !quantum.reg
      }
      scf.yield %13 : !quantum.reg
    }
    %9 = stablehlo.constant dense<0> : tensor<i64>
    %extracted_0 = tensor.extract %9[] : tensor<i64>
    %10 = quantum.extract %8[%extracted_0] : !quantum.reg -> !quantum.bit
    %11 = quantum.namedobs %10[ PauliZ] : !quantum.obs
    %12 = quantum.expval %11 : f64
    %from_elements = tensor.from_elements %12 : tensor<f64>
    quantum.dealloc %7 : !quantum.reg
    return %from_elements : tensor<f64>
  }
```
All of these conditional branches are preserved in the compiled code for future execution of this method.